tests/code_update.py

```diff
@@ -66,22 +75,18 @@ def run(args):
             assert new_node
             new_nodes.add(new_node)
 
-        for node in new_nodes:
-            new_primary = node
-            break
-
         LOG.info("Stopping all original nodes")
         old_nodes = set(network.nodes).difference(new_nodes)
         for node in old_nodes:
             LOG.debug(f"Stopping old node {node.node_id}")
             node.stop()
 
-        LOG.info("Waiting for a new primary to be elected...")
         sleep_time = (
-            args.pbft_view_change_timeout * 6 / 1000
+            args.pbft_view_change_timeout * 2 / 1000
             if args.consensus == "pbft"
-            else args.raft_election_timeout * 6 / 1000
+            else args.raft_election_timeout * 2 / 1000
```

Comment:
Its possible this should still be `3` rather than `2`, I'm not sure. We really want to wait for a successful election + everybody has heard about it, but if we're still seeing election failures this needs to be `2 * timeout + gossip_time` rather than `1 * timeout + gossip_time`

Reply:
gossip time is probably sig-max-ms effectively.

Reply:
Another way to go here, instead of trying to guess the sleep time for each consensus etc, would be to loop a couple of times and wait for the `getPrimary` rpc to respond. That way the test time will depend on how quickly the primary was elected, and the test might also be more stable for the edge cases when the primary took a bit longer to get elected.

Reply:
FYI, we've still got the issue of the first election not completing as the node-to-node connection between the then-follower-now-candidate nodes has not yet been established (vote requests are replaced with n2n connection establishment messages). This is because we do not want to have n^2 n2n connections open at any given times but only n (to the current leader to the followers). The consequence of this is that it takes 2 election rounds (i.e. two timeouts) for a new leader to be elected. 

It might be worth creating an issue for this, at least for documentation purposes?

Reply:
Thanks for the explanation! I've added #1015 to track this.